modmail/utils/pagination.py

```diff
@@ -0,0 +1,260 @@
+    ) -> None:
+        """Create a paginator, and paginate the provided lines."""
+        paginator = cls(
+            contents,
+            source_message=source_message,
+            timeout=timeout,
+            embed=embed,
+            prefix=prefix,
+            suffix=suffix,
+            max_size=max_size,
+            linesep=linesep,
+            only_users=only_users,
+        )
+
+        if channel is None and source_message is None:
+            raise MissingAttributeError("Both channel and message are None.")
+        elif channel is None:
+            channel = source_message.channel
+
+        paginator.modify_states()
+        if len(paginator.pages) >= 2:
+            msg: discord.Message = await channel.send(
+                content=paginator.pages[paginator._index], view=paginator
+            )
+        else:
+            await channel.send(content=paginator.pages[paginator._index])
+            return
+
+        await paginator.wait()
+        await msg.edit(view=None)
+
+    async def interaction_check(self, interaction: "Interaction") -> bool:
+        """Check if the interaction is by the author of the paginatior."""
+        if self.source_message is None:
+            return True
+        if not (is_valid := self.source_message.author.id == interaction.user.id):
+            await interaction.response.send_message(
+                content="This is not your message to paginate!", ephemeral=True
+            )
+        return is_valid
+
+    def modify_states(self) -> None:
+        """Disable specific components depending on paginator page and length."""
+        less_than_2_pages = len(self.pages) <= 2
+        components = {
+            "jump_first": less_than_2_pages,
+            "back": False,
+            "next": False,
+            "jump_last": less_than_2_pages,
+        }
+
+        if self._index == 0:
+            components["jump_first"] = True
+            components["back"] = True
+
+        if self._index == len(self.pages) - 1:
+            components["next"] = True
+            components["jump_last"] = True
+
+        for child in self.children:
+            if child.custom_id in components.keys():
+                if getattr(child, "disabled", None) is not None:
+                    child.disabled = components[child.custom_id]
+                    if getattr(child, "style", None) is not None:
+                        child.style = ButtonStyle.secondary if child.disabled else ButtonStyle.primary
+
+    async def send_page(self, interaction: "Interaction") -> None:
+        """Send new page."""
+        self.modify_states()
+        await interaction.message.edit(content=self.pages[self._index], view=self)
+
+    @button(label=JUMP_FIRST_LABEL, custom_id="jump_first", style=ButtonStyle.primary)
+    async def go_first(self, button: "Button", interaction: "Interaction") -> None:
+        """Move the paginator to the first page."""
+        if self._index == 0:
+            return
```

Comment:
Do we need to double-check here if we are already restricting interactions in the `modify_states` function if I get functionality correctly?

Reply:
No, I don't think so.

Reply:
Actually, if those if statements run, something has seriously broken, and their running will only serve to make it worse.